File: 2017/day16.py
# @formatter:off
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day16.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file(session_path=['..', '.env'])
# @formatter:on

from utils import parsefile

raw_moves = parsefile(file_name, [str, ","])

moves = []
for m in raw_moves:
	if m[0] == "s":  # spin
		moves.append([0, int(m[1:])])
	elif m[0] == "x":  # exchange
		split = m[1:].split("/")
		moves.append([1, int(split[0]), int(split[1])])
	elif m[0] == "p":  # partner
		moves.append([2, ord(m[1]) - 97, ord(m[3]) - 97])

start_order = "abcdefghijklmnop"

from collections import deque


def dance(programs, moves):
	for m in moves:
		if m[0] == 0:  # spin
			programs.rotate(m[1])
		elif m[0] == 1:  # exchange
			temp = programs[m[1]]
			programs[m[1]] = programs[m[2]]
			programs[m[2]] = temp
		elif m[0] == 2:  # partner
			i1 = programs.index(m[1])
			i2 = programs.index(m[2])

			temp = programs[i1]
			programs[i1] = programs[i2]
			programs[i2] = temp


def part1():
	programs = deque([ord(c) - 97 for c in start_order])

	dance(programs, moves)

	return "".join([chr(v + 97) for v in programs])


def part2():
	programs = deque([ord(c) - 97 for c in start_order])

	reset_i = 1

	for i in range(1_000_000_000):
		dance(programs, moves)
		s = "".join([chr(v + 97) for v in programs])
		if s == start_order:
			reset_i = i + 1
			break

	steps_after_start = 1_000_000_000 % reset_i

	for i in range(steps_after_start):
		dance(programs, moves)

	return "".join([chr(v + 97) for v in programs])


p1()
p2()

File: cpp/2017/day16.cpp
#include "../aocHelper.h"

class Day16 : public BaseDay
{
public:
	Day16() : BaseDay("16") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		enum class move_type
		{
			spin,
			exchange,
			partner,
		};

		struct move
		{
			move_type type;
			int a;
			int b;
		};

		vector<move> moves;

		// parse input
		while (*input != '\n')
		{
			if (*input == 's') // spin
			{
				input++; // skip 's'

				int a = numericParse<int>(input);

				moves.push_back({ move_type::spin, a, 0 });
			}
			else if (*input == 'x') // exchange
			{
				input++; // skip 'x'

				int a = numericParse<int>(input);
				input++;
				int b = numericParse<int>(input);

				moves.push_back({ move_type::exchange, a, b });
			}
			else if (*input == 'p') // partner
			{
				input++; // skip 'p'

				int a = *input - 'a';
				input += 2;
				int b = *input - 'a';
				input++;

				moves.push_back({ move_type::partner, a, b });
			}

			if (*input == ',')
			{
				input++; // skip ','
			}
		}

		auto dance = [](array<int, 16>& programs, vector<move>& moves)
		{
			for (auto& m : moves)
			{
				switch (m.type)
				{
					case move_type::spin:
					{
						array<int, 16> tmp_programs;
						for (int i = 0; i < 16; i++)
						{
							tmp_programs[(i + m.a) % 16] = programs[i];
						}

						copy(tmp_programs.begin(), tmp_programs.end(), programs.begin());

						break;
					}
					case move_type::exchange:
					{
						int temp = programs[m.a];
						programs[m.a] = programs[m.b];
						programs[m.b] = temp;

						break;
					}
					case move_type::partner:
					{
						int i1 = 0;
						for (int i = 0; i < 16; i++)
						{
							if (programs[i] == m.a)
							{
								i1 = i;
								break;
							}
						}

						int i2 = 0;
						for (int i = 0; i < 16; i++)
						{
							if (programs[i] == m.b)
							{
								i2 = i;
								break;
							}
						}

						int temp = programs[i1];
						programs[i1] = programs[i2];
						programs[i2] = temp;

						break;
					}
				}
			}
		};

		array<int, 16> programs;
		for (int i = 0; i < 16; i++)
		{
			programs[i] = i;
		}

		// part 1
		array<int, 16> p1_programs;
		copy(programs.begin(), programs.end(), p1_programs.begin());

		dance(p1_programs, moves);

		for (int i = 0; i < 16; i++)
		{
			stringResult.first[i] = (char) (p1_programs[i] + 'a');
		}

		// part 2
		array<int, 16> p2_programs;
		copy(programs.begin(), programs.end(), p2_programs.begin());

		int reset_i = 1;

		for (int i = 0; i < 1'000'000'000; i++)
		{
			dance(p2_programs, moves);

			if (p2_programs == programs)
			{
				reset_i = i + 1;
				break;
			}
		}

		int steps_after_start = 1'000'000'000 % reset_i;

		for (int i = 0; i < steps_after_start; i++)
		{
			dance(p2_programs, moves);
		}

		for (int i = 0; i < 16; i++)
		{
			stringResult.second[i] = (char) (p2_programs[i] + 'a');
		}

		return { part1, part2 };
	}
};
